stellargraph/data/unsupervised_sampler.py

```diff
@@ -141,49 +160,62 @@ def generator(self, batch_size):
         while not done:
             self.random.shuffle(self.nodes)
             for node in self.nodes:  # iterate over root nodes
-                # Get 1 walk at a time. For now its assumed that its a uniform random walker
-                walk = self.walker.run(
-                    nodes=[node],  # root nodes
-                    length=self.length,  # maximum length of a random walk
-                    n=1,  # number of random walks per root node
-                )
-
-                # (target,contect) pair sampling - GraphSAGE way
+                # Get 1 walk at a time. For now its assumed that its a uniform random walker or biased random walker
+                if self.context_sampling:
+                    walk = self.walker.run(
+                        nodes=[node],  # root nodes
+                        length=int(
+                            np.ceil(self.length * self.random.random())
+                        ),  # maximum length of a random walk
+                        n=1,  # number of random walks per root node
+                    )
+                else:
+                    walk = self.walker.run(
+                        nodes=[node],  # root nodes
+                        length=self.length,  # maximum length of a random walk
+                        n=1,  # number of random walks per root node
+                    )
+                # (target,contect) pair sampling
                 target = walk[0][0]
                 context_window = walk[0][1:]
                 for context in context_window:
                     # Don't add self pairs
                     if context == target:
                         continue
 
-                    positive_pairs.append((target, context))
-                    sample_counter += 1
+                    for _ in range(2):
```

Comment:
why is this loop needed now?

Reply:
Here the bidirectinoal=True can increase the diversity of the node context pairs in a same batch. If bidirectional is set to false, the node context pairs generated in a same batch tend to share a comment target nodes. When the option is set to Ture, we will have node context pairs with different target nodes. This is helpful in my previous tests.